_posts/2019-08-09-pairwise-distance.ipynb

```diff
@@ -0,0 +1,720 @@
+      "|   0  Tesla V100-SXM2...  On   | 00000000:06:00.0 Off |                    0 |\n",
+      "| N/A   36C    P0    57W / 300W |   1365MiB / 32510MiB |      0%      Default |\n",
+      "+-------------------------------+----------------------+----------------------+\n"
+     ]
+    }
+   ],
+   "source": [
+    "!nvidia-smi | head"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from numba import cuda"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "How long to move the data?"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "CPU times: user 8 ms, sys: 0 ns, total: 8 ms\n",
+      "Wall time: 56.3 ms\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "<numba.cuda.cudadrv.devicearray.DeviceNDArray at 0x7ff01ec0f940>"
+      ]
+     },
+     "execution_count": 19,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "%%time\n",
+    "x_cuda = cuda.to_device(x)\n",
+    "x_cuda"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import math\n",
+    "\n",
+    "\n",
+    "@cuda.jit(device=True)\n",
+    "def square_coords_cuda(pair_index, n):\n",
+    "    pair_index = np.float32(pair_index)\n",
+    "    n = np.float32(n)\n",
+    "    j = (((2 * n) - 1) - math.sqrt((1 - (2 * n)) ** 2 - (8 * pair_index))) // 2\n",
+    "    k = pair_index - (j * ((2 * n) - j - 1) / 2) + j + 1\n",
+    "    j = np.int64(j)\n",
+    "    k = np.int64(k)\n",
+    "    return j, k\n",
+    "\n",
+    "\n",
+    "@cuda.jit\n",
+    "def kernel_cityblock_cuda(x, out):\n",
+    "    m = x.shape[0]\n",
+    "    n = x.shape[1]\n",
+    "    n_pairs = (n * (n - 1)) // 2\n",
+    "    pair_index = cuda.grid(1)\n",
+    "    if pair_index < n_pairs:\n",
+    "        # Unpack the pair index to column indices.\n",
+    "        j, k = square_coords_cuda(pair_index, n)\n",
+    "        # Iterate over rows, accumulating distance.\n",
+    "        d = np.float32(0)\n",
+    "        for i in range(m):\n",
+    "            u = np.float32(x[i, j])\n",
+    "            v = np.float32(x[i, k])\n",
+    "            d += math.fabs(u - v)\n",
+    "        # Store distance result.\n",
+    "        out[pair_index] = d\n",
+    "\n",
+    "        \n",
+    "def pairwise_cityblock_cuda(x):\n",
+    "\n",
+    "    # Set up output array.\n",
+    "    n = x.shape[1]\n",
+    "    n_pairs = (n * (n - 1)) // 2\n",
+    "    out = cuda.device_array(n_pairs, dtype=np.float32)\n",
+    "\n",
+    "    # Let numba decide number of threads and blocks.\n",
+    "    kernel_spec = kernel_cityblock_cuda.forall(n_pairs)\n",
```

Comment:
Here we use the raveled index to pass to `forall` (`n_pairs` is computed a few lines earlier).